lib/THC/THCGeneral.c

```diff
@@ -17,22 +17,81 @@ void THCudaInit(THCState* state)
   state->blasState = (THCBlasState*)malloc(sizeof(THCBlasState));
   THCudaBlas_init(state, count, device);
 
+  state->numDevices = count;
   state->deviceProperties =
     (struct cudaDeviceProp*)malloc(count * sizeof(struct cudaDeviceProp));
 
-  int i,j;
-  for(i=0; i < count; ++i)
+  state->numUserStreams = 0;
+  state->streamsPerDevice =
+    (cudaStream_t**)malloc(count * sizeof(cudaStream_t*));
+
+  /* Enable P2P access between all pairs, if possible */
+  THCudaEnablePeerToPeerAccess(state);
+
+  for (int i = 0; i < count; ++i)
   {
     THCudaCheck(cudaSetDevice(i));
     THCudaCheck(cudaGetDeviceProperties(&state->deviceProperties[i], i));
+    /* Stream index 0 will be the default stream for convenience; by
+       default no user streams are reserved */
+    state->streamsPerDevice[i] =
+      (cudaStream_t*)malloc(sizeof(cudaStream_t));
+    state->streamsPerDevice[i][0] = NULL;
+  }
 
-    for (j=0; j < count; ++j)
-    {
-      if(i != j)
-      {
+  /* Restore to previous device */
+  THCudaCheck(cudaSetDevice(device));
+
+  /* Start in the default stream on the current device */
+  state->currentPerDeviceStream = 0;
+  state->currentStream = NULL;
+}
+
+void THCudaShutdown(THCState* state)
+{
+  THCRandom_shutdown(state);
+  free(state->blasState);
```

Comment:
We shouldn't free `state->blasState` before calling `THCudaBlas_shutdown`.

I know this code wasn't changed in this PR, but since it's being touched anyway, I thought we might as well fix it now.